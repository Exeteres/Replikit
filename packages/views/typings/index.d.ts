export * from "./viewProps";
export * from "./viewAction";

export * from "./extensions/router";
