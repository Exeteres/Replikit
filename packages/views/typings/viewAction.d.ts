export interface ViewAction {
    name: string;
    arguments: unknown[];
}
