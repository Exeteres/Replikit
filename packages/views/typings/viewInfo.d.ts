import { SafeFunction } from "@replikit/core/typings";
import { OutMessageLikeAsync } from "@replikit/messages/typings";
import { ViewField } from "@replikit/views";

export type ViewRenderer = () => OutMessageLikeAsync;

export interface ViewInfoPrototype {
    [key: string]: SafeFunction;
    render: ViewRenderer;
}

export interface ViewInfo {
    fields: ViewField[];
    prototype: ViewInfoPrototype;
}
