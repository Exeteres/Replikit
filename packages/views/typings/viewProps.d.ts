import { FilterKeysNot, Identifier } from "@replikit/core/typings";
import { View } from "@replikit/views";

export type PropField<T = unknown, R extends boolean = boolean> = T & {
    __propRequired: R;
};

export type InferPropType<P> = P extends PropField<infer T> ? T : P;

type ViewPropFields<T> = {
    [K in keyof T]: T[K] extends PropField
        ? T[K]["__propRequired"] extends true
            ? InferPropType<T[K]>
            : InferPropType<T[K]> | undefined
        : never;
};

type PickNotNever<T> = Pick<T, FilterKeysNot<T, never | undefined>>;

export interface ViewTarget {
    controller: string;
    accountId: Identifier;
}

export type ViewProps<T extends View> = Partial<PickNotNever<ViewPropFields<T>>>;
