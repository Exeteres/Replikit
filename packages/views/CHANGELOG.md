# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.22.1](https://github.com/Exeteres/Replikit/compare/v0.22.0...v0.22.1) (2020-10-23)

**Note:** Version bump only for package @replikit/views





# [0.22.0](https://github.com/Exeteres/Replikit/compare/v0.21.0...v0.22.0) (2020-10-23)

**Note:** Version bump only for package @replikit/views





# [0.21.0](https://github.com/Exeteres/Replikit/compare/v0.20.0...v0.21.0) (2020-10-17)


### Features

* **views:** Add text fallback mechanism and authentication ([d6e926d](https://github.com/Exeteres/Replikit/commit/d6e926d98a9fc10c18db52dee6d8145fae2d14bd))





# [0.20.0](https://github.com/Exeteres/Replikit/compare/v0.19.0...v0.20.0) (2020-09-28)


### Features

* **views:** Update to use composition system ([30733ba](https://github.com/Exeteres/Replikit/commit/30733ba119d99eba8a2085ae8c5f2681d00bd9f6))





# [0.19.0](https://github.com/Exeteres/Replikit/compare/v0.18.2...v0.19.0) (2020-09-18)


### Features

* **views:** Add experimental view system ([8874b49](https://github.com/Exeteres/Replikit/commit/8874b493152ff228b645f25baa8a034ea4383795))
