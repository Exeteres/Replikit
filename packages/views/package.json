{
    "name": "@replikit/views",
    "version": "0.19.0",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "tsc -p tsconfig.build.json"
    },
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@replikit/core": "^0.19.0",
        "@replikit/router": "^0.19.0",
        "@replikit/sessions": "^0.19.0"
    }
}
