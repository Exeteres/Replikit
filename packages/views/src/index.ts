/// <reference path="../typings/index.d.ts" />

export * from "./sessions/viewSession";

export * from "./viewMessageBuilder";
export * from "./startup";
export * from "./utils";
export * from "./errors";
export * from "./view";
export * from "./view";
export * from "./viewStorage";
export * from "./viewField";

export * from "./extensions/messageContextExtension";
