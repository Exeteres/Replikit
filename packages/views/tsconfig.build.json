{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/views": ["views/src"]
        }
    },
    "include": ["src"]
}
