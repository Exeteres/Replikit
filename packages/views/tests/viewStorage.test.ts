import { OutMessage } from "@replikit/core/typings";
import { fromText } from "@replikit/messages";
import { prop, state, View, ViewAlreadyRegisteredError, ViewStorage } from "@replikit/views";

class TestView extends View {
    startFrom = prop(0);
    count = state(this.startFrom);

    increment(): void {
        this.count++;
        this.update();
    }

    render(): OutMessage {
        return fromText(`Count: ${this.count}`);
    }
}

describe("viewStorage", () => {
    it("should register a view and collect its composition info", () => {
        const viewStorage = new ViewStorage();
        viewStorage.register(TestView);
        expect(viewStorage._viewMap).toMatchSnapshot();
    });

    it("should throw an error if view with the same name already registered", () => {
        const viewStorage = new ViewStorage();
        viewStorage.register(TestView);
        expect(() => viewStorage.register(TestView)).toThrow(ViewAlreadyRegisteredError);
    });

    it("should create a view by name", () => {
        const viewStorage = new ViewStorage();
        viewStorage.register(TestView);
        const view = viewStorage.resolve(undefined!, "TestView", { messageIds: [456] });
        expect(view).toMatchSnapshot();
    });
});
