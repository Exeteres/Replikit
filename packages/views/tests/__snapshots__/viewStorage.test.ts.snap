// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewStorage should create a view by name 1`] = `
TestView {
  "_data": Object {
    "count": 0,
    "startFrom": 0,
  },
  "metadata": Object {
    "messageIds": Array [
      456,
    ],
  },
}
`;

exports[`viewStorage should register a view and collect its composition info 1`] = `
Map {
  "TestView" => Object {
    "fields": Object {},
    "prototype": TestView {
      "__init__": [Function],
    },
  },
}
`;
